{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "type object 'HTMLAwareEntitySubstitution' has no attribute 'preserve_whitespace_tags'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-3f36240f4087>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpprint\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mbs4\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0msubprocess\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/bs4/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     33\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mwarnings\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 35\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mbuilder\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mbuilder_registry\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mParserRejectedMarkup\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     36\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mdammit\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mUnicodeDammit\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m from .element import (\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/bs4/builder/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    226\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    227\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 228\u001b[0;31m \u001b[0;32mclass\u001b[0m \u001b[0mHTMLTreeBuilder\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mTreeBuilder\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    229\u001b[0m     \"\"\"This TreeBuilder knows facts about HTML.\n\u001b[1;32m    230\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/bs4/builder/__init__.py\u001b[0m in \u001b[0;36mHTMLTreeBuilder\u001b[0;34m()\u001b[0m\n\u001b[1;32m    232\u001b[0m     \"\"\"\n\u001b[1;32m    233\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 234\u001b[0;31m     \u001b[0mpreserve_whitespace_tags\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mHTMLAwareEntitySubstitution\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpreserve_whitespace_tags\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    235\u001b[0m     empty_element_tags = set([\n\u001b[1;32m    236\u001b[0m         \u001b[0;31m# These are from HTML5.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: type object 'HTMLAwareEntitySubstitution' has no attribute 'preserve_whitespace_tags'"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "import xlrd\n",
    "import re\n",
    "import requests\n",
    "import json\n",
    "import pprint\n",
    "import bs4\n",
    "import time\n",
    "import subprocess\n",
    "from Bio import AlignIO\n",
    "import jinja2\n",
    "import urllib.request, urllib.error, urllib.parse\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "\n",
    "class homolog_finder():\n",
    "    def __init__(self, xls_filename=\"\",\n",
    "                 urlUniprotId='http://www.uniprot.org/uniprot/?query=id',\n",
    "                 urlCompara='https://rest.ensembl.org/homology/symbol/caenorhabditis_elegans/',\n",
    "                 urlBrain='http://api.brain-map.org/api/v2/data/query.json?criteria=',\n",
    "                 urlWormbase='http://www.wormbase.org/rest/widget',\n",
    "                 urlOverlap='http://rest.ensembl.org/overlap',\n",
    "                 urlUniprot='http://www.uniprot.org/uniprot',\n",
    "                 urlDisgenet='http://www.disgenet.org/oql',\n",
    "                 main_folder='',\n",
    "                 wormbase_name_column=4,\n",
    "                 wormbase_id_column=2,\n",
    "                 uniprot_id_column=3,\n",
    "                 sprot_database='uniprot_sprot.fasta',\n",
    "                 jackhmmer='jackhmmer',\n",
    "                 disease_ontology_filename='HumanDO.obo'):\n",
    "        self.xls_filename = xls_filename\n",
    "        self.urls = dict()\n",
    "        self.urls['xrefs'] = 'http://rest.ensembl.org/xrefs/id'\n",
    "        self.urls['id'] = urlUniprotId\n",
    "        self.urls['compara'] = urlCompara\n",
    "        self.urls['brain'] = urlBrain\n",
    "        self.urls['wormbase'] = urlWormbase\n",
    "        self.urls['overlap'] = urlOverlap\n",
    "        self.urls['uniprot'] = urlUniprot\n",
    "        self.urls['disgenet'] = urlDisgenet\n",
    "        self.required_folders = ['brain', 'compara', 'wormbase_protein', 'aliases', 'disgenet', 'uniprot']\n",
    "        self.main_folder = main_folder\n",
    "        self.disgenet = dict() #stores all the gene to disease assocations\n",
    "        self.disgenet_raw = dict() #stores all raw data from disgenet\n",
    "        self.disgenet = dict() #stores parsed data from disgenet\n",
    "        self.gene_ids = list() #stores a list of gene IDs for which homologs are searched\n",
    "        self.uniprot_ids = list() #tores a list of UniProt IDs for which homologs are searched\n",
    "        self.initialize_folders()\n",
    "        self.primary_genes = list()\n",
    "        self.homolog_gene = list()\n",
    "        self.homologs = dict() #key: Uniprot ID or gene name, value: a list of homolog object\n",
    "        self.sprot_database = sprot_database #location of the SwissProt database for HMMER\n",
    "        self.jackhmmer = jackhmmer #location of the jackhmmer executable\n",
    "        self.uniprot_genename = dict() #key: UniProt ID, value: Gene name\n",
    "        self.expressed_in_brain = dict() #key: gene name, value: boolean (expressed in brain or not)\n",
    "        self.wormbase_names = list() #a list of wormbase names\n",
    "        self.neuro_diseases = list() #a list of all diseases associated with brain or nerves\n",
    "        self.disease_synonyms = dict() #a dict with synonyms for neuro_diseases, key: a disease ontology entry, value: the corresponding MESH value\n",
    "        self.disease_ontology_filename = disease_ontology_filename #the filename where the human disease ontology is stored\n",
    "        self.uniprot_info = dict() #stores info about Uniprot entries which were retrieved, key: Uniprot ID, value: a dict with information\n",
    "        self.disgenet_map = None #a placeholder for a dict, key: UniProtID, value: GeneID\n",
    "        self.disgenet_map_filename ='mapa_geneid_4_uniprot_crossref.tsv'\n",
    "        \n",
    "        if xls_filename:\n",
    "            #if a xls filename is specified, read all the data\n",
    "            #initalize primary genes for all entries and find possible homologs\n",
    "            self.workbook = xlrd.open_workbook(self.xls_filename)\n",
    "            #if not self.validate_inputfile():\n",
    "            #    sys.exit(1)\n",
    "            self.get_genes_from_xls_file(wormbase_id_column)\n",
    "            self.gene_ids = self.get_ids_from_xls_file(wormbase_id_column)\n",
    "            self.uniprot_ids = self.get_ids_from_xls_file(uniprot_id_column)\n",
    "            self.gene_ids = self.gene_ids\n",
    "            self.wormbase_names = self.get_ids_from_xls_file(wormbase_name_column)\n",
    "            if len(self.gene_ids) != len(self.uniprot_ids):\n",
    "                print('Gene length and Uniprot length do not match: {0} vs {1}'.format(len(self.gene_ids), len(self.uniprot_ids)))\n",
    "                sys.exit(1)\n",
    "            self.find_homologs()\n",
    "        else:\n",
    "            self.workbook = None\n",
    "\n",
    "\n",
    "        #parse and store the disease ontology\n",
    "        if self.disease_ontology_filename:\n",
    "            import pronto\n",
    "            onto = pronto.Ontology(self.disease_ontology_filename)\n",
    "            parents = [onto['DOID:150'], onto['DOID:863']] #mental and nerve diseases\n",
    "            for parent in parents:\n",
    "                for c, child in enumerate(parent.children):\n",
    "                    self.neuro_diseases.append(child)\n",
    "                    while child.children:\n",
    "                        child = child.children\n",
    "                        self.neuro_diseases.extend(child)\n",
    "\n",
    "            for neuro_disease in self.neuro_diseases:\n",
    "                if neuro_disease.other.get('xref'):\n",
    "                    for term in neuro_disease.other['xref']:\n",
    "                        if term.startswith('MESH:'):\n",
    "                            self.disease_synonyms[term.split('MESH:')[1]] = neuro_disease\n",
    "\n",
    "    def calculate_scores(self):\n",
    "        \"\"\"\n",
    "        \n",
    "        :return: \n",
    "        \"\"\"\n",
    "\n",
    "\n",
    "        for primary_gene in self.homologs:\n",
    "            for homolog in self.homologs[primary_gene]:\n",
    "                homolog.calculate_score(homolog_finder=self)\n",
    "\n",
    "        return None\n",
    "\n",
    "    def get_homologs(self):\n",
    "        \"\"\"\n",
    "        \n",
    "        :return: \n",
    "        \"\"\"\n",
    "        return list(self.homologs.values())[0]\n",
    "\n",
    "    def print_homologs(self):\n",
    "        \"\"\"\n",
    "        prints information about homologs\n",
    "        :return: None\n",
    "        \"\"\"\n",
    "\n",
    "        for primary_gene in self.homologs:\n",
    "            print('Primary gene: {}'.format(primary_gene))\n",
    "            for homolog in self.homologs[primary_gene]:\n",
    "                print(homolog)\n",
    "                print('=' * 20)\n",
    "\n",
    "    def get_human_homologs(self):\n",
    "        \"\"\"\n",
    "        returns all human homologs\n",
    "        :return: \n",
    "        \"\"\"\n",
    "\n",
    "        self.human_homologs = dict()\n",
    "        for primary_gene in self.homologs:\n",
    "            print('#' * 40)\n",
    "            print('Primary gene: {}'.format(primary_gene))\n",
    "            self.human_homologs[primary_gene] = list()\n",
    "            for homolog in self.homologs[primary_gene]:\n",
    "                if 'homo' in homolog.organism.lower() or 'human' in homolog.organism.lower():\n",
    "                    self.human_homologs[primary_gene].append(homolog)\n",
    "                    print(homolog)\n",
    "                    print('=' * 20)\n",
    "\n",
    "        for primary_gene in self.human_homologs:\n",
    "            print('Primary gene {} has {} potential homologs'.format(primary_gene,\n",
    "                                                                     len(self.human_homologs[primary_gene])\n",
    "                                                                     )\n",
    "                  )\n",
    "\n",
    "    def get_info_from_uniprot(self, uniprot_id):\n",
    "        \"\"\"\n",
    "        retrieves information from Uniprot and stores it in uniprot_info attribute\n",
    "        :param uniprot_id: a Uniprot ID, e.g. Q9TU53\n",
    "        :return: a dictionary with attributes\n",
    "        \"\"\"\n",
    "\n",
    "        attributes = dict()\n",
    "        attributes['organism'] = ''\n",
    "        attributes['fullname'] = ''\n",
    "        attributes['sequence'] = ''\n",
    "        attributes['gene_id'] = ''\n",
    "        filename = 'uniprot/{}.xml'.format(uniprot_id)\n",
    "        if not os.path.isfile(filename):\n",
    "            r = requests.get('{0}/{1}.xml'.format(self.urls['uniprot'], uniprot_id))\n",
    "            if r.status_code == 200:\n",
    "                with open(filename, 'w') as f:\n",
    "                    f.write(r.text)\n",
    "            else:\n",
    "                print('Error retrieving Uniprot {}'.format(uniprot_id))\n",
    "                return attributes\n",
    "\n",
    "        #with open(filename, 'r') as f:\n",
    "        #    uniprot_xml = f.read()\n",
    "\n",
    "        root = ET.parse(filename).getroot()\n",
    "        entries = [node for node in root.getchildren() if 'entry' in node.tag]\n",
    "        if len(entries) != 1:\n",
    "            print('malformed XML for UniProt ID: {}\\nexpected 1 entry'.format(uniprot_id))\n",
    "\n",
    "\n",
    "        organism_root = [node for node in entries[0].getchildren() if node.tag.endswith('organism')]\n",
    "        if len(organism_root) != 1:\n",
    "            print('malformed XML for UniProt ID: {}\\nexpected 1 organism'.format(uniprot_id))\n",
    "        else:\n",
    "            organism = [node for node in organism_root[0].getchildren() if\n",
    "                    (node.tag.endswith('name') and node.attrib.get('type') == 'common')]\n",
    "            if len(organism) == 1:\n",
    "                attributes['organism'] = organism[0].text\n",
    "            else:\n",
    "                organism = [node for node in organism_root[0].getchildren() if\n",
    "                        (node.tag.endswith('name') and node.attrib.get('type') == 'scientific')]\n",
    "                if len(organism) == 1:\n",
    "                    attributes['organism'] = organism[0].text\n",
    "                else:\n",
    "                    print('malformed XML for UniProt ID: {}\\nexpected 1 common organism name'.format(uniprot_id))\n",
    "\n",
    "        seq = [node for node in root[0].getchildren() if node.tag.endswith('sequence')]\n",
    "        if len(seq) != 1:\n",
    "            print('malformed XML for UniProt ID: {}\\nexptected 1 sequence'.format(uniprot_id))\n",
    "        else:\n",
    "            attributes['sequence'] = seq[0].text.strip().replace('\\n', '')\n",
    "\n",
    "\n",
    "        protein = [node for node in root[0].getchildren() if node.tag.endswith('protein')]\n",
    "        if len(protein) != 1:\n",
    "            print('malformed XML for UniProt ID: {}\\nexptected 1 protein'.format(uniprot_id))\n",
    "            attributes['fullname'] = ''\n",
    "        else:\n",
    "            submittedName = [node for node in protein[0].getchildren() if node.tag.endswith('submittedName')]\n",
    "            if len(submittedName) != 1:\n",
    "                submittedName = [node for node in protein[0].getchildren() if node.tag.endswith('recommendedName')]\n",
    "            if len(submittedName) != 1:\n",
    "                print('malformed XML for UniProt ID: {}\\nexpected 1 submittedName or recommendedName'.format(uniprot_id))\n",
    "            else:\n",
    "                attributes['fullname'] = submittedName[0].getchildren()[0].text\n",
    "\n",
    "        gene_id_root = [node for node in entries[0].getchildren() if node.tag.endswith('gene')]\n",
    "        gene_found = False\n",
    "        if len(gene_id_root) != 1:\n",
    "            print('malformed XML for UniProt ID: {}\\nexpected 1 gene\\nTrying via API'.format(uniprot_id))\n",
    "        else:\n",
    "            #print(gene_id_root)\n",
    "            gene_id = [node for node in gene_id_root[0].getchildren() if\n",
    "                       (node.tag.endswith('name') and node.attrib.get('type') == 'primary')]\n",
    "            if len(gene_id) == 1:\n",
    "                gene_found = True\n",
    "                attributes['gene_id'] = gene_id[0].text\n",
    "        if not gene_found:\n",
    "            print('Could not find GeneName in XML for UniProt ID: {}\\nTrying via API'.format(uniprot_id))\n",
    "            attributes['gene_id'] = self.uniprot_to_genename(uniprot_id)\n",
    "\n",
    "        return attributes\n",
    "    def find_homologs(self):\n",
    "        for ii, gene in enumerate(self.gene_ids):\n",
    "            self.primary_genes.append(primary_gene(uniprot_id=self.uniprot_ids[ii],\n",
    "                                                   wormbase_id=gene,\n",
    "                                                   homolog_finder=self,\n",
    "                                                   wormbase_name=self.wormbase_names[ii]))\n",
    "            self.primary_genes[-1].find_homologs()\n",
    "        for primary in self.primary_genes:\n",
    "            for uniprot in primary.uniprot_ids_from_hmmer:\n",
    "                self.create_homolog(primary,\n",
    "                                    uniprot=uniprot,\n",
    "                                    source='hmmer')\n",
    "            for compara in primary.compara:\n",
    "                self.create_homolog(primary,\n",
    "                                    gene_name=compara,\n",
    "                                    source='compara')\n",
    "            self.create_homolog_html(primary)\n",
    "\n",
    "    def get_homolog(self, uniprot_id):\n",
    "        return self.homologs.get(uniprot_id)\n",
    "\n",
    "    def create_homolog(self,\n",
    "                       primary,\n",
    "                       gene_name='',\n",
    "                       uniprot='',\n",
    "                       source='',\n",
    "                       double_check=True):\n",
    "        \"\"\"\n",
    "        \n",
    "        :param primary: \n",
    "        :param double_check: \n",
    "        :return: \n",
    "        \"\"\"\n",
    "        if double_check and self.get_homolog(primary.wormbase_name):\n",
    "            if uniprot and uniprot in self.get_homolog(primary.wormbase_name):\n",
    "                return None\n",
    "            if gene_name and gene_name in self.get_homolog(primary.wormbase_name):\n",
    "                return None\n",
    "\n",
    "        new_homolog = homolog_gene(primary,\n",
    "                                   uniprot_id=uniprot,\n",
    "                                   gene_name=gene_name,\n",
    "                                   source=source,\n",
    "                                   homolog_finder=self\n",
    "                                   )\n",
    "        if not self.homologs.get(primary.wormbase_name):\n",
    "            self.homologs[primary.wormbase_name] = list()\n",
    "        self.homologs[primary.wormbase_name].append(new_homolog)\n",
    "\n",
    "    def initialize_folders(self):\n",
    "        \"\"\"\n",
    "        creates all input folders needed to run the script\n",
    "        :return: None\n",
    "        \"\"\"\n",
    "\n",
    "        for folder in self.required_folders:\n",
    "            joinedFolder = os.path.join(folder, self.main_folder)\n",
    "            if not os.path.isdir(joinedFolder):\n",
    "                os.makedirs(joinedFolder)\n",
    "\n",
    "    def validate_inputfile(self):\n",
    "        \"\"\"\n",
    "        validates the input file\n",
    "        :return:\n",
    "        \"\"\"\n",
    "        # let's check if the gene matches the ID, UniProt and name\n",
    "        xl_sheet = self.workbook.sheet_by_index(0)\n",
    "        wb_genes = xl_sheet.col(0)\n",
    "        print('validating genes')\n",
    "        for i, wb_gene in enumerate(wb_genes):\n",
    "            error = (None, '')\n",
    "            print('Validating {0}'.format(wb_gene.value))\n",
    "            r = requests.get(\n",
    "                '{0}/{1}?content-type=application/json'.format(self.urls['xrefs'], wb_gene.value.strip(), timeout=10))\n",
    "            if r.status_code == 200:\n",
    "                data = json.loads(r.text)\n",
    "            else:\n",
    "                print('error retrieving data for {0}'.format(wb_gene))\n",
    "                data = list()\n",
    "            # UniProt check\n",
    "            for d in data:\n",
    "                if 'UniProt' in d.get('db_display_name'):\n",
    "                    if not xl_sheet.cell_value(rowx=i, colx=3) == d.get('primary_id'):\n",
    "                        if error[0] == None:\n",
    "                            error = (True, d.get('primary_id'))\n",
    "                    else:\n",
    "                        error = (False, '')\n",
    "            if error[0] == True:\n",
    "                print('Error in row {0}: Given Uniprot ID {1} does not match retrieved Uniprot ID {2}'.format(i + 1,\n",
    "                                                                                                              xl_sheet.cell_value(\n",
    "                                                                                                                  rowx=i,\n",
    "                                                                                                                  colx=3),\n",
    "                                                                                                              error[1]))\n",
    "\n",
    "            if error[0] == None:\n",
    "                print('Warning in row {0}: No Uniprot ID found'.format(i + 1))\n",
    "            # ID check\n",
    "            error = (None, '')\n",
    "            for d in data:\n",
    "                if 'wormbase_gseqname' in d.get('dbname'):\n",
    "                    if not xl_sheet.cell_value(rowx=i, colx=2) == d.get('display_id'):\n",
    "                        if error[0] == None:\n",
    "                            error = (True, d.get('display_id'))\n",
    "                    else:\n",
    "                        error = (False, '')\n",
    "            if error[0] == True:\n",
    "                print('Error in row {0}: Given display ID {1} does not match retrieved display ID {2}'.format(i + 1,\n",
    "                                                                                                              xl_sheet.cell_value(\n",
    "                                                                                                                  rowx=i,\n",
    "                                                                                                                  colx=2),\n",
    "                                                                                                              error[1]))\n",
    "\n",
    "            if error[0] == None:\n",
    "                print('Warning in row {0}: No Uniprot ID found'.format(i + 1))\n",
    "\n",
    "\n",
    "            # let's check if the gene matches the ID, UniProt and name\n",
    "\n",
    "    def get_genes_from_xls_file(self, col):\n",
    "        \"\"\"\n",
    "        Reads the gene IDs from the Excel file\n",
    "        updates the gene_ids attribute\n",
    "        :param col: \n",
    "        :return: a list of gene IDs\n",
    "        \"\"\"\n",
    "\n",
    "        xl_sheet = self.workbook.sheet_by_index(0)\n",
    "        cols = xl_sheet.col(col)\n",
    "\n",
    "        p = re.compile('^.*\\(([-.a-zA-Z0-9]+)\\)')\n",
    "\n",
    "        for c in cols:\n",
    "            m = p.match(c.value)\n",
    "            if m:\n",
    "                self.gene_ids.append(m.group(1))\n",
    "            else:\n",
    "                self.gene_ids.append('')\n",
    "        return self.gene_ids\n",
    "\n",
    "    def get_ids_from_xls_file(self, col):\n",
    "        \"\"\" \n",
    "        :param col: \n",
    "        :return: \n",
    "        \"\"\"\n",
    "        xl_sheet = self.workbook.sheet_by_index(0)\n",
    "        cols = xl_sheet.col(col)\n",
    "\n",
    "        ids = [c.value for c in cols]\n",
    "        return ids\n",
    "\n",
    "    def render(self, template_path, context):\n",
    "        \"\"\"\n",
    "    \n",
    "        :param template_path: location of the template\n",
    "        :param context: a dictionary which is used for the template\n",
    "        :return: the parsed template as a string\n",
    "        \"\"\"\n",
    "        path, filename = os.path.split(template_path)\n",
    "        return jinja2.Environment(\n",
    "            loader=jinja2.FileSystemLoader(path or './')\n",
    "        ).get_template(filename).render(context)\n",
    "\n",
    "    def create_homolog_html(self, primary):\n",
    "        \"\"\"\n",
    "        \n",
    "        :param primary: \n",
    "        :return: \n",
    "        \"\"\"\n",
    "        input_dict = dict()\n",
    "        input_dict['primary'] = primary.wormbase_name\n",
    "        input_dict['homologs'] = self.homologs[primary.wormbase_name]\n",
    "        table = list()\n",
    "        table.append(list())\n",
    "        table[0].append('Uniprot ID')\n",
    "        table[0].append('Source')\n",
    "        table[0].append('Expressed in brain')\n",
    "        table[0].append('Alignment')\n",
    "        filename = 'jackhmmer_{}.aln'.format(primary.uniprot_id)\n",
    "        for homolog in self.homologs[primary.wormbase_name]:\n",
    "            table.append(list())\n",
    "            table[-1].append('<a href=\"http://www.uniprot.org/uniprot/{0}\">{0}</a>'.format(homolog.uniprot_id))\n",
    "            table[-1].append(homolog.source)\n",
    "            if homolog.expressed_in_brain:\n",
    "                table[-1].append('yes')\n",
    "            elif homolog.expressed_in_brain == False:\n",
    "                table[-1].append('no')\n",
    "            else:\n",
    "                table[-1].append('unknown')\n",
    "            table[-1].append('<a href=\"{}.html\" target=\"_blank\">Open alignment</a>'.format(filename))\n",
    "        result = self.render('table.template.html', {'table': table})\n",
    "        print(primary.wormbase_name)\n",
    "        with open('{}.html'.format(primary.wormbase_name.replace('/', '_')), 'w') as f:\n",
    "            f.write(result)\n",
    "        self.alignment(filename)\n",
    "\n",
    "\n",
    "    def alignment(self, filename='align.txt'):\n",
    "        \"\"\"\n",
    "        Creates an alignment file from an alignment, only human sequences are used\n",
    "        :param filename: the file where the alignment is saved\n",
    "        :return: None\n",
    "        \"\"\"\n",
    "        alignment = AlignIO.read(filename, \"stockholm\")\n",
    "        items = list()\n",
    "        for i, a in enumerate(alignment):\n",
    "            if 'HUMAN' in a.id or i == 0:\n",
    "                items.append({'id': a.id, 'sequence': str(a.seq)})\n",
    "\n",
    "        result = self.render('alignment.template', {'items': items})\n",
    "\n",
    "        with open('{}.html'.format(filename), 'w') as f:\n",
    "            f.write(result)\n",
    "\n",
    "\n",
    "    def hmmer_to_table(self, filename='h.txt'):\n",
    "\n",
    "        #the location of the individual rows\n",
    "        ranges = [(0, 22), (22, 33), (33, 56), (56, 66), (67, 76), (77, 83), (84, 89), (90, 99), (100, 106), (107, 112), (114, 118), (118, 122), (122, 126), (126, 130), (130, 134), (134, 138), (138, 142), (142, 146), (146, 10**3)]\n",
    "        lines = list()\n",
    "        header = 0\n",
    "        with open(filename, 'r') as f:\n",
    "            for line in f:\n",
    "                if line.startswith('#'):\n",
    "                    header += 1\n",
    "                if header == 2 or not line.startswith('#'):\n",
    "                    lines.append(list())\n",
    "                    for r in ranges:\n",
    "                        lines[-1].append(line[r[0]:r[1]].strip())\n",
    "                if header > 3:\n",
    "                    break\n",
    "        lines[0][1] = 'Expressed in the brain'\n",
    "        for i, line in enumerate(lines[1:]):\n",
    "            gene = self.uniprot_to_genename(line[0].split('|')[1])\n",
    "            lines[i + 1][1] = expressed_in_brain(gene)\n",
    "\n",
    "        result = self.render('table.template.html', {'table': lines})\n",
    "        with open('table.html', 'w') as f:\n",
    "            f.write(result)\n",
    "\n",
    "\n",
    "    def uniprot_to_genename(self, uniprot):\n",
    "        \"\"\"\n",
    "        gets the gene name for a UniProt identifier\n",
    "        :param uniprot: a UniProt identifier\n",
    "        :return: a gene name, in case of multiple gene names only the first is returned\n",
    "        \"\"\"\n",
    "\n",
    "        if not self.uniprot_genename.get(uniprot):\n",
    "            self.uniprot_genename[uniprot] = None\n",
    "            filename = 'uniprot/id_{}.txt'.format(uniprot)\n",
    "            if not os.path.isfile(filename):\n",
    "                r = requests.get('{0}:{1}&columns=id,entry%20name,genes&format=json'.format(self.urls['id'],\n",
    "                                                                                        uniprot))\n",
    "\n",
    "                if r.status_code == 200:\n",
    "                    with open(filename, 'w') as f:\n",
    "                        f.write(r.text)\n",
    "                    self.uniprot_genename[uniprot] = json.loads(r.text)[0]['genes'].split(' ')[0]\n",
    "            else:\n",
    "                with open(filename, 'r') as f:\n",
    "                    self.uniprot_genename[uniprot] =json.load(f)[0]['genes'].split(' ')[0]\n",
    "\n",
    "        return self.uniprot_genename[uniprot].split(';')[0]\n",
    "\n",
    "    def get_diseases_for_genes(self):\n",
    "        \"\"\"\n",
    "        \n",
    "        :return: \n",
    "        \"\"\"\n",
    "        for gene in self.get_homologs():\n",
    "            self.get_disease_for_gene(self, gene.gene_id)\n",
    "\n",
    "\n",
    "    def get_disease_for_gene(self, gene):\n",
    "        \"\"\"\n",
    "        \n",
    "        :param gene: a string with GeneID, e.g. SEZ6 \n",
    "        :return: \n",
    "        \"\"\"\n",
    "\n",
    "\n",
    "    def read_disgenet_map(self):\n",
    "        \"\"\"\n",
    "        \n",
    "        :return: \n",
    "        \"\"\"\n",
    "\n",
    "        with open(self.disgenet_map_filename, 'r') as f:\n",
    "            lines = f.read().splitlines()\n",
    "\n",
    "        self.disgenet_map = dict()\n",
    "        for line in lines:\n",
    "            cells = line.split('\\t')\n",
    "            self.disgenet_map[cells[0]] = cells[1]\n",
    "\n",
    "        return None\n",
    "    def get_diseases_from_disgenet(self, uniprot_id, filename='', overwrite=False):\n",
    "        \"\"\"\n",
    "\n",
    "        inspired by:\n",
    "        http://www.disgenet.org/ds/DisGeNET/scripts/disgenet_python3.py\n",
    "        :param gene:\n",
    "        :return:\n",
    "        \"\"\"\n",
    "\n",
    "        if not self.disgenet_map:\n",
    "            self.read_disgenet_map()\n",
    "        gene_id = self.disgenet_map.get(uniprot_id)\n",
    "        if not gene_id:\n",
    "            print('no gene ID found for {}'.format(uniprot_id))\n",
    "            return None\n",
    "        if not filename:\n",
    "            filename = 'disgenet/{}'.format(uniprot_id)\n",
    "        if not overwrite:\n",
    "            if os.path.isfile(filename):\n",
    "                with open(filename, 'r') as f:\n",
    "                    return (f.read())\n",
    "\n",
    "\n",
    "#        query = \"\"\"\n",
    "#                DEFINE\n",
    "#                    c0='/data/gene_disease_summary',\n",
    "#            c1='/data/diseases',\n",
    "#            c2='/data/genes',\n",
    "#            c3='/data/gene_roles',\n",
    "#            c4='/data/sources'\n",
    "#                ON\n",
    "#                   'http://www.disgenet.org/web/DisGeNET'\n",
    "#                SELECT\n",
    "#                    c2 (geneId, name, uniprotId, description, pathName, pantherName),\n",
    "#                    c1 (cui, name, diseaseClassName, STY, MESH, omimInt),\n",
    "#                   c3 (PI, PL),\n",
    "#                  c0 (score, pmids,  snps)\n",
    "#\n",
    "#                FROM\n",
    "#                    c0\n",
    "#                WHERE\n",
    "#                    (\n",
    "#                        c2.name = '{0}'\n",
    "#                    AND\n",
    "#                        c4 = 'ALL'\n",
    "#                    )\n",
    "#                ORDER BY\n",
    "#                    c0.score DESC\"\"\".format(uniprot_id)\n",
    "        query = \"\"\"\n",
    "        DEFINE\n",
    "          \tc0='/data/gene_disease_summary',\n",
    "\tc1='/data/diseases',\n",
    "\tc2='/data/genes',\n",
    "\tc4='/data/sources'\n",
    "        ON\n",
    "           'http://www.disgenet.org/web/DisGeNET'\n",
    "        SELECT\n",
    "         \tc1 (diseaseId, name, diseaseClassName, STY, MESH, OMIM, type ),\n",
    "\tc2 (geneId, symbol,   uniprotId, description, pantherName ),\n",
    "\tc0 (score, EI, Npmids, Nsnps)\n",
    "           \n",
    "        FROM\n",
    "            c0\n",
    "        WHERE\n",
    "            (\n",
    "                c2.geneId = '{0}'\n",
    "            AND\n",
    "                c4 = 'ALL'\n",
    "            )\n",
    "        ORDER BY\n",
    "            c0.score DESC\"\"\".format(gene_id)\n",
    "        query = query.encode('utf-8')\n",
    "        req = urllib.request.Request(self.urls['disgenet'])\n",
    "        res = urllib.request.urlopen(req, query)\n",
    "        data = res.read().decode(\"utf-8\")\n",
    "        if overwrite or not os.path.isfile(filename):\n",
    "            with open(filename, 'w') as f:\n",
    "                f.write(data)\n",
    "        #self.parse_disgenet(data)\n",
    "        return (data)\n",
    "\n",
    "    def get_disgenet(self, uniprot_id):\n",
    "        \"\"\"\n",
    "\n",
    "        :param uniprot_id: The uniprot ID for which associated diseases should be found\n",
    "        :return:\n",
    "        \"\"\"\n",
    "        if not self.disgenet.get(uniprot_id):\n",
    "            filename = 'disgenet/{}.txt'.format(uniprot_id)\n",
    "            data = self.get_diseases_from_disgenet(uniprot_id, filename)\n",
    "            self.disgenet[uniprot_id] = self.parse_disgenet(self.filter_disgenet(data))\n",
    "\n",
    "        return self.disgenet[uniprot_id]\n",
    "\n",
    "    def filter_disgenet(self, data):\n",
    "        \"\"\"\n",
    "\n",
    "        :param self: \n",
    "        :param data: a result string from a DisGenet query \n",
    "        :return: all lines which are associated with nerve disease, taken from self.neuro_diseases\n",
    "        \"\"\"\n",
    "        filtered_diseases = list()\n",
    "        if not data:\n",
    "            return filtered_diseases\n",
    "\n",
    "        lines = data.split('\\n')\n",
    "        for line in lines:\n",
    "            cells = line.split('\\t')\n",
    "            if len(cells) < 11:\n",
    "                continue\n",
    "            if self.disease_synonyms.get(cells[10]):\n",
    "                filtered_diseases.append(line)\n",
    "\n",
    "        return filtered_diseases\n",
    "\n",
    "    def parse_disgenet(self, data):\n",
    "        \"\"\"\n",
    "        \n",
    "        :param data: \n",
    "        :return: \n",
    "        \"\"\"\n",
    "\n",
    "        parsed_values = dict()\n",
    "        parsed_values['total score'] = 0\n",
    "        parsed_values['diseases'] = list()\n",
    "        for line in data:\n",
    "            cells = line.split('\\t')\n",
    "            parsed_values['total score'] += float(cells[14])\n",
    "            parsed_values['diseases'].append(dict(disease=cells[7], score=float(cells[14])))\n",
    "        return parsed_values\n",
    "\n",
    "class primary_gene():\n",
    "    def __init__(self,\n",
    "                 uniprot_id='',\n",
    "                 wormbase_id='',\n",
    "                 protein_name='',\n",
    "                 wormbase_name='',\n",
    "                 homolog_finder=None,\n",
    "                 overwrite=False):\n",
    "        self.uniprot = uniprot_id\n",
    "        self.wormbase_id = wormbase_id\n",
    "        self.wormbase_name = wormbase_name\n",
    "        self.protein_name = protein_name\n",
    "        self.uniprot_id = uniprot_id\n",
    "        self.homolog_finder = homolog_finder #the outer homolog_finder object which stores all the settings\n",
    "        self.homologs = list() #stores a list of homolog objects\n",
    "        self.compara = list() #stores the info from compara database\n",
    "        self.overwrite = False #whether to overwrite existing data or not\n",
    "        self.jackhmmer_filename = '' #the filename where the jackhmmer output is located\n",
    "        self.wormbase_homologs = ''\n",
    "        self.overwrite = overwrite\n",
    "        self.uniprot_ids_from_hmmer = list()\n",
    "        if homolog_finder and uniprot_id:\n",
    "            self.gene_id = homolog_finder.uniprot_to_genename(uniprot_id)\n",
    "        else:\n",
    "            self.gene_id = ''\n",
    "\n",
    "    def find_homologs(self):\n",
    "        self.get_homologs_from_compara()\n",
    "        self.run_jackhmmer()\n",
    "        self.get_uniprot_ids_from_hmmer()\n",
    "        self.get_info_from_wormbase()\n",
    "\n",
    "    def run_jackhmmer(self):\n",
    "        \"\"\"\n",
    "        \n",
    "        :return: \n",
    "        \"\"\"\n",
    "        filename = '{0}.fasta'.format(self.uniprot_id)\n",
    "        jackhmmer_output = 'jackhmmer_{0}.aln'.format(self.uniprot_id)\n",
    "\n",
    "        if not os.path.isfile(filename):\n",
    "            r = requests.get('{0}/{1}.fasta'.format(self.homolog_finder.urls['uniprot'], self.uniprot_id))\n",
    "            with open(filename, 'w') as f:\n",
    "                f.write(r.text)\n",
    "        if not os.path.isfile(jackhmmer_output) or self.overwrite:\n",
    "            args = [self.homolog_finder.jackhmmer,\n",
    "                    '-A',\n",
    "                    jackhmmer_output,\n",
    "                    filename,\n",
    "                    self.homolog_finder.sprot_database]\n",
    "            subprocess.check_output(args)\n",
    "        self.jackhmmer_filename = jackhmmer_output\n",
    "\n",
    "    def get_uniprot_ids_from_hmmer(self):\n",
    "        \"\"\"\n",
    "        gets all UniProt IDs from a hmmer alignment file\n",
    "        does not return the original input ID\n",
    "        :param hmmer_file: the filename which is searched\n",
    "        :return: a list of IDs in order of appearance\n",
    "        \"\"\"\n",
    "\n",
    "        ids = list()\n",
    "        if not os.path.isfile(self.jackhmmer_filename):\n",
    "            self.uniprot_ids_from_hmmer = list()\n",
    "            return False\n",
    "\n",
    "        with open(self.jackhmmer_filename, 'r') as f:\n",
    "            lines = f.read().split('\\n')\n",
    "        if not lines[0].startswith('# STOCKHOLM'):\n",
    "            self.uniprot_ids_from_hmmer = list()\n",
    "            return False\n",
    "\n",
    "        orig_id = lines[1].split('|')[1]\n",
    "        for line in lines[3:]:\n",
    "            if line.startswith('#'):\n",
    "                id = line.split('|')[1]\n",
    "                if id not in ids:\n",
    "                    ids.append(id)\n",
    "            else:\n",
    "                break\n",
    "        self.uniprot_ids_from_hmmer = ids\n",
    "        return True\n",
    "\n",
    "    def calculate_score_from_alignment(self):\n",
    "        \"\"\"\n",
    "        calculates the score of the individual alignments\n",
    "        :return: \n",
    "        \"\"\"\n",
    "\n",
    "    def get_homologs_from_compara(self,\n",
    "                                  species='homo_sapiens'):\n",
    "        filename = 'compara/{0}_{1}.txt'.format(species, self.wormbase_id)\n",
    "        if not os.path.isfile(filename):\n",
    "            r = requests.get(\n",
    "                '{0}{1}?content-type=application/json&target_species={2}'.format(self.homolog_finder.urls['compara'],\n",
    "                                                                                 self.wormbase_id,\n",
    "                                                                                 species))\n",
    "            data = r.text\n",
    "            with open(filename, 'w') as f:\n",
    "                f.write(data)\n",
    "        else:\n",
    "            print(filename)\n",
    "            with open(filename, 'r') as f:\n",
    "                data = f.read()\n",
    "\n",
    "        j = json.loads(data)\n",
    "        if not 'error' in j.keys() and len(j['data']) > 0:\n",
    "            for i in range(len(j['data'][0]['homologies'])):\n",
    "                self.compara.append(j['data'][0]['homologies'][i]['target']['id'])\n",
    "\n",
    "    def get_info_from_wormbase(self):\n",
    "        filename = 'wormbase_protein/{0}.html'.format(self.wormbase_id)\n",
    "\n",
    "        if not os.path.isfile(filename):\n",
    "            r = requests.get('{0}/cds/{1}/overview'.format(self.homolog_finder.urls['wormbase'], self.wormbase_id))\n",
    "            if r.status_code != 200:\n",
    "                r = requests.get('http://www.wormbase.org/rest/widget/cds/{0}a/overview'.format(self.wormbase_id))\n",
    "            data = r.text.encode('ascii', errors='ignore')\n",
    "            with open(filename, 'wb') as f:\n",
    "                f.write(data)\n",
    "        else:\n",
    "            with open(filename, 'rb') as f:\n",
    "                data = f.read()\n",
    "\n",
    "        filename = 'wormbase_protein/{0}_protein.html'.format(self.wormbase_id)\n",
    "        if not os.path.isfile(filename):\n",
    "            soup = bs4.BeautifulSoup(data)\n",
    "            links = soup.find_all('a', {'class': 'protein-link'})\n",
    "            if links:\n",
    "                r = requests.get(\n",
    "                    '{0}/protein/{1}/overview'.format(self.homolog_finder.urls['wormbase'], links[0].get('href').split('/')[-1]))\n",
    "                data = r.text\n",
    "            else:\n",
    "                data = ''\n",
    "            with open(filename, 'w') as f:\n",
    "                f.write(data)\n",
    "        else:\n",
    "            with open(filename, 'r') as f:\n",
    "                data = f.read()\n",
    "        soup = bs4.BeautifulSoup(data)\n",
    "        links = soup.find_all('a', {'class': 'protein-link'})\n",
    "        if links:\n",
    "            self.wormbase_homologs = [links[0].get('href').split(':')[-1], links[0].text]\n",
    "            return True\n",
    "        else:\n",
    "            self.wormbase_homologs = ['', '']\n",
    "            return False\n",
    "\n",
    "\n",
    "class homolog_gene():\n",
    "    \"\"\"\n",
    "    stores all the information about homologs\n",
    "    \"\"\"\n",
    "    def __init__(self,\n",
    "                 primary, #the origin of the homology, a primary gene object\n",
    "                 uniprot_id='', #the uniprot ID, used as primary identifier\n",
    "                 gene_name='', #the gene name, used as primary identifier\n",
    "                 jackhmmer_files='', #the filenames with the jackhmmer alignment\n",
    "                 jackhmmer_ids=list(), #a list of uniprot IDs identified by jackhmmer\n",
    "                 source='', #from where the homology was deduced, e.g. hmmer or compara\n",
    "                 homolog_finder=None,\n",
    "                 score=None, #the best (for multiple homologs) weighted score taking alignment, expression, etc. into account\n",
    "                 description='', #description, taken from UniProt\n",
    "                 name='', #human readable name of the homolog\n",
    "                 organism='',\n",
    "                 sequence=''\n",
    "                 ):\n",
    "        self.primary = primary\n",
    "        self.uniprot_id = uniprot_id\n",
    "        self.score = score\n",
    "        self.source = source\n",
    "        self.description = description\n",
    "        self.name = name\n",
    "        self.organism = organism\n",
    "        self.sequence = sequence\n",
    "        self.disgenet = None #a place holder for info from disgenet\n",
    "        if type(jackhmmer_files) == list:\n",
    "            self.jackhmmer_files = jackhmmer_files\n",
    "        else:\n",
    "            self.jackhmmer_files = list()\n",
    "            self.jackhmmer_files.append(jackhmmer_files)\n",
    "\n",
    "        self.jackhmmer_ids = jackhmmer_ids\n",
    "        if not gene_name and homolog_finder and uniprot_id:\n",
    "            self.gene_name = homolog_finder.uniprot_to_genename(uniprot_id)\n",
    "        else:\n",
    "            self.gene_name = gene_name.upper()\n",
    "        if self.gene_name:\n",
    "            self.expressed_in_brain = self.expressed_in_brain(self.gene_name)\n",
    "        else:\n",
    "            self.expressed_in_brain = None\n",
    "\n",
    "        if uniprot_id and homolog_finder:\n",
    "            uniprot_info = homolog_finder.get_info_from_uniprot(uniprot_id)\n",
    "            #print(uniprot_info)\n",
    "            if not self.name:\n",
    "                self.name = uniprot_info['fullname']\n",
    "            #if not self.description:\n",
    "            #    self.description = uniprot_info['description']\n",
    "            if not self.sequence:\n",
    "                self.sequence = uniprot_info['sequence']\n",
    "            if not self.organism:\n",
    "                self.organism = uniprot_info['organism']\n",
    "\n",
    "    def __repr__(self):\n",
    "        return 'Primary: {0}\\nName: {1}\\nUniprot: {2}\\nOrganism: {3}\\nScore: {4}\\nExpressed in Brain: {5}\\n'.format(self.primary.wormbase_name, self.name, self.uniprot_id, self.organism, self.score, self.expressed_in_brain)\n",
    "\n",
    "    def add_info(self,\n",
    "                 jackhmmer_file=None,\n",
    "                 jackhmmer_ids=None,\n",
    "                 compara=None,\n",
    "                 primary=None):\n",
    "\n",
    "        if primary:\n",
    "            jackhmmer_ids = primary.uniprot_ids_from_hmmer\n",
    "            jackhmmer_file = primary.jackhmmer_filename\n",
    "            compara = primary.compara\n",
    "        if jackhmmer_file:\n",
    "            if jackhmmer_file not in self.jackhmmer_files:\n",
    "                self.jackhmmer_files.append(jackhmmer_file)\n",
    "        if jackhmmer_ids:\n",
    "            if type(jackhmmer_ids) != list:\n",
    "                jackhmmer_ids = [jackhmmer_ids]\n",
    "            for jackhmmer_id in jackhmmer_ids:\n",
    "                if jackhmmer_id not in self.jackhmmer_ids:\n",
    "                    self.jackhmmer_ids.append(jackhmmer_id)\n",
    "        if compara:\n",
    "            if type(compara) != list:\n",
    "                compara = list(compara)\n",
    "            for comp in compara:\n",
    "                if comp not in self.compara:\n",
    "                    self.compara.append(comp)\n",
    "\n",
    "    def expressed_in_brain(self,\n",
    "                           gene,\n",
    "                           donor='15496',\n",
    "                           thresh_avg=3,\n",
    "                           thresh_max=5):\n",
    "        \"\"\"Checks if a gene is expressed in the brain via the Allen Brain Atlas\n",
    "        :param gene: a gene name according to HGNC\n",
    "        :param donor: the id of the brain donor\n",
    "        :param thresh_avg: threshold for average in all measurements\n",
    "        :param thresh_max: threshold which needs to met at least once\n",
    "        :return: True, if above thresholds, False, if below thresholds\n",
    "        \"\"\"\n",
    "        gene = gene.upper()\n",
    "        exp_max = 0\n",
    "        exp_avg = 0\n",
    "        probes = list()\n",
    "\n",
    "        filename = 'brain/{0}_probes.txt'.format(gene.replace('/', '_'))\n",
    "        if os.path.isfile(filename):\n",
    "            with open(filename, 'r') as f:\n",
    "                msg = json.load(f)\n",
    "        else:\n",
    "            r = requests.get(\n",
    "                \"http://api.brain-map.org/api/v2/data/query.json?criteria=model::Probe,rma::criteria,[probe_type$eq'DNA'],products[abbreviation$eq'HumanMA'],gene[acronym$eq'{0}'],rma::options[only$eq'probes.id']\".format(\n",
    "                    gene),\n",
    "                timeout=10)\n",
    "\n",
    "            if r.status_code != 200:\n",
    "                return None\n",
    "            msg = json.loads(r.text)['msg']\n",
    "            with open(filename, 'w') as f:\n",
    "                json.dump(msg, f)\n",
    "        for id in msg:\n",
    "            probes.append(str(id['id']))\n",
    "        filename = 'brain/{0}_expression_donor_{1}.txt'.format(gene.replace('/', '_'), donor)\n",
    "        if os.path.isfile(filename):\n",
    "            with open(filename, 'r') as f:\n",
    "                try:\n",
    "                    probes = json.load(f)\n",
    "                except:\n",
    "                    return None\n",
    "        else:\n",
    "            r = requests.get(\n",
    "                \"http://api.brain-map.org/api/v2/data/query.json?criteria=service::human_microarray_expression[probes$eq{0}][donors$eq{1}]\".format(\n",
    "                    ','.join(probes), donor), timeout=10)\n",
    "            if r.status_code == 200:\n",
    "                print(r.text)\n",
    "                if json.loads(r.text).get('success'):\n",
    "                    probes = json.loads(r.text)['msg']['probes']\n",
    "                    with open(filename, 'w') as f:\n",
    "                        json.dump(probes, f)\n",
    "                else:\n",
    "                    with open(filename, 'w') as f:\n",
    "                        f.write('')\n",
    "                    return None\n",
    "            else:\n",
    "                return None\n",
    "        for probe in probes:\n",
    "            exp_level = [float(p) for p in probe['expression_level']]\n",
    "            exp_max += max(exp_level)\n",
    "            exp_avg += sum(exp_level) / len(exp_level)\n",
    "\n",
    "        exp_max /= len(probes)\n",
    "        exp_avg /= len(probes)\n",
    "\n",
    "        if exp_avg > thresh_avg and exp_max > thresh_max:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "\n",
    "    def calculate_score(self,\n",
    "                        weights=dict(brain=dict(values=[-1, 0, 1],\n",
    "                                                weight=1),\n",
    "                                     hmmer=dict(weight=1)),\n",
    "                        homolog_finder=None):\n",
    "        values = (False, None, True)\n",
    "        if self.score == None:\n",
    "            if self.expressed_in_brain in values:\n",
    "                self.score = weights['brain']['values'][values.index(self.expressed_in_brain)]\n",
    "                self.score *= weights['brain']['weight']\n",
    "            if not self.disgenet and self.organism.lower() in ('human', 'homo sapiens'):\n",
    "                self.disgenet = homolog_finder.get_disgenet(self.uniprot_id)\n",
    "            else:\n",
    "                self.disgenet = dict()\n",
    "                self.score += 0 #nothing found, explicitly not increasing score\n",
    "\n",
    "        return self.score\n",
    "\n",
    "\n",
    "\n",
    "#new_homolog_finder = homolog_finder(xls_filename='57_top_candidates.xlsx')\n",
    "new_homolog_finder = homolog_finder(xls_filename='57_top_candidates.xlsx')\n",
    "#for k in new_homolog_finder.homologs.keys():\n",
    "#    pass\n",
    "\n",
    "\n",
    "\n",
    "values = dict()\n",
    "for i in range(1, 10):\n",
    "    values[str(i)] = i\n",
    "values['-'] = 0\n",
    "values['.'] = 0\n",
    "values['*'] = 10\n",
    "\n",
    "#for i in range(1, len(alignment)):\n",
    "#    print(alignment[i].letter_annotations['posterior_probability'].count('*') / len(alignment[0].seq))\n",
    "\n",
    "k = list(new_homolog_finder.homologs.keys())[0]\n",
    "print(new_homolog_finder.homologs[k][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "type object 'HTMLAwareEntitySubstitution' has no attribute 'preserve_whitespace_tags'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-f05ae67a88d7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpprint\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mbs4\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0msubprocess\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/bs4/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     33\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mwarnings\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 35\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mbuilder\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mbuilder_registry\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mParserRejectedMarkup\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     36\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mdammit\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mUnicodeDammit\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m from .element import (\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/bs4/builder/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    226\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    227\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 228\u001b[0;31m \u001b[0;32mclass\u001b[0m \u001b[0mHTMLTreeBuilder\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mTreeBuilder\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    229\u001b[0m     \"\"\"This TreeBuilder knows facts about HTML.\n\u001b[1;32m    230\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/bs4/builder/__init__.py\u001b[0m in \u001b[0;36mHTMLTreeBuilder\u001b[0;34m()\u001b[0m\n\u001b[1;32m    232\u001b[0m     \"\"\"\n\u001b[1;32m    233\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 234\u001b[0;31m     \u001b[0mpreserve_whitespace_tags\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mHTMLAwareEntitySubstitution\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpreserve_whitespace_tags\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    235\u001b[0m     empty_element_tags = set([\n\u001b[1;32m    236\u001b[0m         \u001b[0;31m# These are from HTML5.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: type object 'HTMLAwareEntitySubstitution' has no attribute 'preserve_whitespace_tags'"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "import xlrd\n",
    "import re\n",
    "import requests\n",
    "import json\n",
    "import pprint\n",
    "import bs4\n",
    "import time\n",
    "import subprocess\n",
    "from Bio import AlignIO\n",
    "import jinja2\n",
    "import urllib.request, urllib.error, urllib.parse\n",
    "import xml.etree.ElementTree as ET"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
